Feature/jjob hofx

ShastriPaturi opened the pull request:
## Description
This PR is for running the SOCA HOFX to be used as an independent verification tool.
It adds a new case hofx3d.yaml to the list of available cases and currently tests only ADT. 
This case assumes that a 24 forecast cycle (on MOM6CICE50.25 model) has been successfully completed. A new environmental variable SOCA_ANALYSIS is created to list the path. 

### Issue(s) addressed
Addresses issue #145 

## Testing
Please test the HOFX for ADT only. Additional QC filters are needed to test for profile TS. The profile TS fails on Oct 02, 2011.
a. In the workflow/cases/hofx3d.yaml file, 
      da_settings: 
      Keep DA_ADT: True and the others false. 
      Please add "/scratch2/NCEPDEV/marineda/Shastri.Paturi/Shastri.Paturi/rundir/workflowtest_QCexp_SST25" to SOCA_ANALYSIS. 
      This has the 24h fcst run from Oct 1-31, 2011 with satellite SST assimilation.

ShastriPaturi committed:
JJB for HOFX only

ShastriPaturi committed:
JJOB for HOFX

ShastriPaturi committed:
Scripts for HOFX

ShastriPaturi committed:
remove JJOB_DA_PREP. This is not necessary for this PR

ShastriPaturi committed:
Add HOFX JJOBS

guillaumevernieres commented on jobs/JJOB_HOFX_PREP:
```diff
@@ -0,0 +1,121 @@
+# Copy SOCA analysis from a previous run to run the HOFX
+# This overwrite the MOM.res.nc from ${SOCA_STATIC}
+
+cp -p /scratch2/NCEPDEV/marineda/Shastri.Paturi/Shastri.Paturi/rundir/workflowtest_QCexp_SST25/$CDATE/Data/ocn.socagodas.an.${bkg_date}.nc $RUNCDATE/INPUT_MOM6/MOM.res.nc
```
Looks fishy:
1. You are pointing to the analysis as opposed to the background
2. What about sea-ice?

ShastriPaturi replied:
1. Isn't the analysis of day 1, the background for day 2 (copied as MOM.res.nc)?
2. Next PR

guillaumevernieres replied:
nope, the 24hrs forecast from the analysis is the background for the next cycle. Not the same thing.

ShastriPaturi replied:
@guillaumevernieres, is INPUT_MOM6/MOM.res.nc, the 24 hr analysis/background?
Is it not over-written after each cycle?

guillaumevernieres replied:
@ShastriPaturi : The background is under fcst when you run the coupled model and I'm afraid it is lost when running with mom6solo. That's a bug, my bad.
I suggest you keep pointing ocn....ana.nc file, since we know what these are. Just add the sea-ice analysis.

guillaumevernieres commented on scripts/prep_hofx3d_yaml.sh:
```diff
@@ -30,12 +30,16 @@ done
 sed -i "s/BKG_DATE/${bkg_date}/g" ${yamlfile}
 sed -i "s/MEMBER_NO/${mbr}/g" ${yamlfile}
+sed -i "s/ocn.pert.ens..nc/MOM.res.nc/g" ${yamlfile}
+sed -i "s/cic.pert.ens..nc/cice_bkg.nc/g" ${yamlfile}
```
why do you need perturbations? Are you looking at an ensemble of hofx?

ShastriPaturi replied:
The perturbations are part of the ensemble runs from @flampouris.
I did not modify it and also. the script needs a 3rd argument "mbr" for ensembles.

@flampouris need your suggestion on this.

flampouris replied:
This copying is happening because the letkf uses the Hofx job. Keep it in, and i will modify it at the next PR.

guillaumevernieres commented on workflow/layout/hofx_only.yaml:
```diff
@@ -0,0 +1,48 @@
+        config: [ base ]
+        J_JOB: JJOB_HOFX_PREP
+
+      da_prep_soca: !Task
```
That one step is over-kill, I guess this is fine for this pr, but we need to document the problem in the issues. This will generate the grid, which is needed, but also compute the bump coef, not needed and resource intensive ~5-10 mn on 400 cores for the 1/4 degree case.

ShastriPaturi replied:
Agreed. Making the change.

flampouris reviewed (changes requested):
@ShastriPaturi very useful contribution. A couple of comments:

1. We still need the JJOB_DA_PREP, please do not delete it. 
2. Clean the commented out code from the scripts.
3. Documentation: 1. what this case is doing and 2. how can be used.
4. No hard coded paths. 

That's all.

ShastriPaturi committed:
Added JJOB_DA_PREP

ShastriPaturi committed:
JJOBS for HOFX

ShastriPaturi committed:
case yaml for hof3d

ShastriPaturi commented:
> ## Description
> 
> This PR is for running the SOCA HOFX to be used as an independent verification tool.
> This PR adds a new case hofx3d.yaml to the list of available cases.
> This PR currently tests only ADT.
> ### Issue(s) addressed
> 
> Addresses issue #145
> ## Testing
> 
> Please test the HOFX for ADT only. Additional QC filters are needed to test for profile TS.
> For example: ./setup_case.sh -p HERA ../cases/hofx3d.yaml workflowtest_HOFX
> In the godas/workflow/cases/hofx3d.yaml, make sure the DA_ADT=True and others are False. Run for two days: Oct 2011, 01 and 02.



> @ShastriPaturi very useful contribution. A couple of comments:
> 
>     1. We still need the JJOB_DA_PREP, please do not delete it.
> 
>     2. Clean the commented out code from the scripts.
> 
>     3. Documentation: 1. what this case is doing and 2. how can be used.
> 
>     4. No hard coded paths.
> 
> 
> That's all.

1. JJOB_DA_PREP added again.
2. Cleaned
3. Added 1. in the description. 2. Added a README file
4. Done

ShastriPaturi committed:
Removed hard-coded path. Added Environmental variable

ShastriPaturi commented:
> @ShastriPaturi very useful contribution. A couple of comments:
> 
>     1. We still need the JJOB_DA_PREP, please do not delete it.
> 
>     2. Clean the commented out code from the scripts.
> 
>     3. Documentation: 1. what this case is doing and 2. how can be used.
> 
>     4. No hard coded paths.
> 
> 
> That's all.

1. JJOB_DA_PREP added again.
2.Cleaned
3.Added
4 Done

flampouris commented on workflow/defaults/default_resources.yaml:
```diff
@@ -36,7 +36,7 @@ godas_resource_table: !Select
       r3denvar:     [  240,   40, !timedelta "01:00:00", 1,   "3072" ]
       da_prep:      [    1,    1, !timedelta "00:30:00", 1,   "3072" ]
-      obs_prep:     [   12,   12, !timedelta "00:10:00", 1,   "3072" ]
+      obs_prep:     [   12,   12, !timedelta "01:30:00", 1,   "3072" ]
```
Why do we need 1.5 hours?

ShastriPaturi replied:
@flampouris, 1.5 hrs was added for the raw data to be pre-processed. for HofX, it is not needed as we already have the pre-processed data.
I can change it to10 mins as previous.

flampouris replied:
I see, there are two issues to consider then:
1. The obs_prep time has to be updated to all the 3 cases.
2. For now, I think that it's better to keep the 10', as you have already preprocessed the obs for 3 months, the obs_prep just copies.

ShastriPaturi replied:
@flampouris changed the obs_prep time for  Ocean025deg to 10 mins.

flampouris commented on workflow/defaults/places.yaml:
```diff
@@ -18,9 +18,11 @@ default_places: &default_places
   SOCA_CONFIG:   !expand "{ROOT_GODAS_DIR}/src/soca-bundle/soca-config/emc"             # soca-config path
   DCOM_ROOT: !expand "{GODAS_RC}/ocean_observations/data"         # Observation root path for raw obs
-  IODA_ROOT: !expand "{GODAS_RC}/ioda_observations"               # Processed observations
+  IODA_ROOT: !expand "{GODAS_RC}/ioda_observations"   # Processed observations
```
Please keep the original code.

ShastriPaturi replied:
@flampouris Ok.

flampouris replied:
This is still unresolved

ShastriPaturi replied:
merged develop and resolved

flampouris commented on workflow/defaults/places.yaml:
```diff
@@ -18,9 +18,11 @@ default_places: &default_places
   SHORT_TERM_TEMP: !calc doc.platform.short_term_temp             # short term storage
   LONG_TERM_TEMP:  !calc doc.platform.long_term_temp              # long term storage
   RUNDIR: !expand "{LONG_TERM_TEMP}/{tools.env('USER')}/rundir/{doc.names.experiment}" # RUN directory
+  SOCA_ANALYSIS:  "/scratch2/NCEPDEV/marineda/Shastri.Paturi/Shastri.Paturi/rundir/workflowtest_QCexp_SST25"   # SOCA ANALYSIS For HOFX
```
@ShastriPaturi Help me here, this points to your path, how different is it from the RUNDIR?

ShastriPaturi replied:
@flampouris it is different from RUNDIR, since it will be the one in which the present HoFX is running and my path, at the moment, is the RUNDATE for a completed 1 month cycle on which the HofX will be run.
The user will have to update it, when he/she wants to use his/her completed run.

flampouris replied:
I think that I get it, but you should document how you have designed the case.

Let's talk about this in person, no fix paths are permitted to this file, maybe you would like to add the SOCA_ANALYSIS path to the hofx_only.yaml?

ShastriPaturi replied:
@flampouris, we need to talk in person to resolve the location of the SOCA_ANALYSIS path.
layout/hofx_only.yaml, cases/hofx3d.yaml are not the right places for the same.

ShastriPaturi replied:
@flampouris after a careful thought, I have removed the SOCA_ANALYSIS variable from default/place.yaml and added it to cases/hofx3d.yaml. This seems to be the appropriate place for it for the time-being, at least. We can brain storm with @guillaumevernieres and @JianKuang-NOAA, if necessary as an update to this PR. 
If you agree, then I will push the changes.

flampouris reviewed (commented):
The README has not being updated, how can I use the new case?

ShastriPaturi commented:
> The README has not being updated, how can I use the new case?

@flampouris the instructions to test is under Testing below the Description of the PR.
I will update the REAME similar to @jkbk2004 once the PR goes through.

flampouris commented:
Till now, only WIP has gone through without README update,s and there are
no exceptions to this. I consider this PR as v1.0 of the hofx case,
therefore README is required.

Without instructions, useful work is lost.

ShastriPaturi committed:
Merge branch 'develop' into feature/JJOB_HOFX
To get the latest README.md file

ShastriPaturi committed:
Update README file

ShastriPaturi commented:
> Till now, only WIP has gone through without README update,s and there are no exceptions to this. I consider this PR as v1.0 of the hofx case, therefore README is required. Without instructions, useful work is lost.

@flampouris fair statement. Updated the README file.

flampouris reviewed (changes requested):
See the comments in the code

ShastriPaturi committed:
Changed the obs_prep time to original - 10 mins

ShastriPaturi committed:
Moved location for SOCA_ANALYSIS Path

ShastriPaturi committed:
Updates to README

flampouris commented on README.md:
```diff
@@ -172,6 +172,21 @@ The default is 1 member to 1 group (deterministic run).
 :x: No implementation planned <br/>
 :soon: Work in progress
 
+# HofX case as a verification tool
```
Add the hofx3d.yaml to the section describing the cases, at line 95.

flampouris commented on README.md:
```diff
@@ -172,6 +172,21 @@ The default is 1 member to 1 group (deterministic run).
+# HofX case as a verification tool
+0. HofX case can be run separately as an independant verification tool.
+
+1. Instructions on how to run this case are as follows:
```
Add the instructions for the hofx3d.yaml under configuration (line 138) and describe what options the user can use: Which observations and setting the path.

flampouris commented on README.md:
```diff
@@ -172,6 +172,21 @@ The default is 1 member to 1 group (deterministic run).
+
+1. Instructions on how to run this case are as follows:
+
+   a. ./setup_case.sh -p HERA ../cases/hofx3d.yaml workflowtest_HOFX
```
This line is not necessary.

flampouris commented on README.md:
```diff
@@ -172,6 +172,21 @@ The default is 1 member to 1 group (deterministic run).
+   a. ./setup_case.sh -p HERA ../cases/hofx3d.yaml workflowtest_HOFX
+
+   b. In the workflow/cases/hofx3d.yaml file, da_settings: DA_ADT: True and the others false. 
+      Please add "/scratch2/NCEPDEV/marineda/Shastri.Paturi/Shastri.Paturi/rundir/workflowtest_QCexp_SST25" to SOCA_ANALYSIS. 
```
This line should be in the instructions on how to test the code for the PR.

ShastriPaturi replied:
Added

flampouris commented on README.md:
```diff
@@ -172,6 +172,21 @@ The default is 1 member to 1 group (deterministic run).
+
+   c. The results of the run will be in RUNCDATE/Data.
+
+   d. A scatter plot between the obs_absolute_dynamic_topography@ObsValue and obs_absolute_dynamic_topography@hofx is one-way of looking at the results.
```
I don't see any calls to plotting routines, how does the scatter plot is created?

ShastriPaturi replied:
@flampouris, I will add a scatter plot routine to the JJOB_POST in the next version, as discussed.

flampouris commented on jobs/JJOB_DA_PREP:
```diff
@@ -234,3 +234,5 @@ if [ $FCSTMODEL = 'MOM6CICE5' ]; then
    ncecat -O -v aicen,vicen,vsnon -u Time $RUNCDATE/INPUT_MOM6/cice_bkg.nc $RUNCDATE/INPUT_MOM6/cice_bkg.nc
    ncks -A -v Time,xaxis_1,yaxis_1,zaxis_1 $RUNCDATE/INPUT_MOM6/ice_model.res.nc $RUNCDATE/INPUT_MOM6/cice_bkg.nc
 fi
+
```
Reverse the changes on this file, the two empty lines are not necessary.

flampouris commented on workflow/config/base.yaml:
```diff
@@ -65,6 +65,8 @@ config_base:
     export NINNER="{doc.da_settings.NINNER}"
 
+    export SOCA_ANALYSIS="{doc.places.SOCA_ANALYSIS}"
+
```
Delete the empty line.

flampouris commented on workflow/defaults/places.yaml:
```diff
@@ -18,9 +18,10 @@ default_places: &default_places
   SHORT_TERM_TEMP: !calc doc.platform.short_term_temp             # short term storage
   LONG_TERM_TEMP:  !calc doc.platform.long_term_temp              # long term storage
   RUNDIR: !expand "{LONG_TERM_TEMP}/{tools.env('USER')}/rundir/{doc.names.experiment}" # RUN directory
+
```
Delete the empty line

ShastriPaturi replied:
merged develop and resolved

flampouris reviewed (changes requested):
@ShastriPaturi good job. 

1. Merge the develop, there is one conflict. 
2. Address all the 9 comments inside the code.

flampouris commented on jobs/JJOB_PREP_SOCA:
```diff
@@ -38,10 +38,3 @@ echo
 mkdir -p bump
 srun -n $NPE ${SOCA_EXEC}/soca_staticbinit.x ./yaml/static_SocaError_init.yml
 
-# Initialize BUMP localization (computes localization weights)
```
This cannot be deleted, because it is used from all the cases.

ShastriPaturi replied:
Thanks for pointing out.
Added.

flampouris commented on workflow/cases/hofx3d.yaml:
```diff
@@ -0,0 +1,18 @@
+    FCSTMODEL: MOM6CICE5
+
+  da_settings:
+    NINNER: 100
```
The NINNER is not required for the HOFX, just deleted from this file.

ShastriPaturi replied:
Deleted

ShastriPaturi committed:
Merge branch 'develop' into feature/JJOB_HOFX

ShastriPaturi committed:
Addreesed merge conflict

ShastriPaturi committed:
Addressed merge conflict

ShastriPaturi committed:
Addressed comments

flampouris commented on README.md:
```diff
@@ -164,6 +168,34 @@ case:
+    # Observation switches
+    DA_SST: False    # Sea surface temperature
+    DA_TS: False     # T & S Insitu profiles 
+    DA_ADT: True    # Absolute dynamic topography
```
For consistency, update the hofx.yaml.

flampouris commented on workflow/layout/hofx_only.yaml:
```diff
@@ -0,0 +1,48 @@
+        <<: *shared_task_template
+        Trigger: !Depend  hofx_run
+        resources: !calc partition.resources.run_fcst_prep
+        rocoto_command: "echo post"
```
Remove the  rocoto_command: "echo post" (see 3dvar case) and check the figures in /Figures

ShastriPaturi committed:
Made Updates

Commit:
Update README.md

Commit:
Updates for the PR

flampouris commented:
@ShastriPaturi please review the updates to the code and run the hofx and 3dvar cases (this PR was breaking the 3dvar).

If the changes do not satisfy you, please update the code and let me know when it is ready for review.

ShastriPaturi commented:
The 3dvar and HofX runs pass. However, the JJOB_POST fails in that, it can't find the fcst directory in the RUNCATE as it gets deleted.
If that is no issue, then please go ahead and merge.

flampouris reviewed (approved):
👍 

It is an issue; the compromise was to remove the JJOB_post from the layout, but it is still in the code. 

The code will not be merged as-is.

ShastriPaturi committed:
Reverted to original code

ShastriPaturi committed:
Reverted to original code

ShastriPaturi committed:
Reverted to original code
